packages/vitest/src/types/coverage.ts

```diff
@@ -53,12 +53,15 @@ export type CoverageReporter =
   | 'text-summary'
   | 'text'
 
-type Provider = 'c8' | 'istanbul' | CoverageProviderModule | undefined
+type BuiltinProviders = 'c8' | 'istanbul'
+type CustomProvider = string & Omit<string, BuiltinProviders>
+type Provider = BuiltinProviders | CustomProvider | undefined
 
 export type CoverageOptions<T extends Provider = Provider> =
-  T extends CoverageProviderModule ? ({ provider: T } & BaseCoverageOptions) :
-    T extends 'istanbul' ? ({ provider: T } & CoverageIstanbulOptions) :
-        ({ provider?: T } & CoverageC8Options)
+  T extends 'istanbul' ? ({ provider: T } & CoverageIstanbulOptions) :
+    T extends 'c8' ? ({ provider: T } & CoverageC8Options) :
+      T extends CustomProvider ? ({ provider: T } & CustomProviderOptions) :
+          ({ provider?: T } & CoverageC8Options)
```

Comment:
This part is getting really tricky. As the `provider` can be a known value of `c8` | `istanbul`,  it can also be just a dynamic `string`. The `Omit<string, BuiltinProviders>` doesn't work anymore. [TS playground](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAQgVwJYBsRIHYAUBOB7ANyQBMocBnGAXhgHIBjADlpgB86lyQBDDAIwQpaoSLADCCLngC2uQiTLUYXHJgDmMAGQwA8tKQgAPCvUAaeMjSY5RUhQB8I6DBsKcSxKnTZ8tspXYJKVlfNzYYBAxSADNMKGJgJ3FCMm41KABJLl4BFB0IdDwMShoAb2UZTOz+QXzCjAAuGD48PBQoXhgAXySYMRScNKgxRjqkIpKYcvJKkbGippMMDR7wZyCQGVc7eeKladnJTeldpowEaT5FHuAoAA8IPBwQGDXkglT03fJDABUYe4gKBRSjbRQ0ME4ezUYAwGD-QHA4iUWicHg1IQwAD8MAAFOUIKE7E1-l0tH0BkMsujct8AJQwBqwuHwgF3IEguhMFg4-EwQnyYmssnafofQbpOYFcbFBlMlkshHspGUDZbImKXkEjU4EndclqkKCsj0xnMhUKvkCvw4LF6kUU8VDKX1ch0xLAegTF5onKCfoYWJqJpiz5Qb77ZnWtxNABEvoxsdMzJm0iqNNq0oWrxwCCgyeZAHpCzAAAIgcgAWnu0HoIGrOHw7krlfhAAtOMo23hBMQYINOLAyE2U7NRlnGjBY9xyMQk8AugBuRJe4ovJgBoMhylfCeTUpRnVxpjzuGp4bj+pxmdzgtw4tlivVh5QOsNpswFvtzvkbu9-vcIOAKNk8o5ptSfp5BOTQgLm+YLsuQA)

I think we'll need to use 100% known values for `provider`, and use additional option for custom providers where they can define the path for the module. 

E.g.

```ts
provider: 'custom',
path: 'custom-provider.ts',
```